text/1201-naked-fns.md

```diff
@@ -71,16 +71,15 @@ address of its caller when called with the C ABI on x86.
 ---
 
 Because the compiler depends on a function prologue and epilogue to maintain
-storage for local variable bindings, it is generally unsafe to write anything
-but inline assembly inside a naked function.  The [LLVM language
-reference](http://llvm.org/docs/LangRef.html#function-attributes) describes this
-feature as having "very system-specific consequences", which the programmer must
-be aware of.
+storage for local variable bindings, including anything but inline assembly
+inside a naked function is not allowed. Refering to the parameters inside of a
```

Comment:
This seems to suggest that you can have a naked function with parameters, but you can't use them. This seems incorrect - parameters will always have an implicit use due to the generated `drop` calls. Shouldn't naked functions be forbidden from taking any parameters?

Reply:
That was my original intention, but passing arguments still makes sense as you can reference them directly from the registers in the assembly.

Reply:
That would require rust to generate a prologue, though - which naked functions are supposed to avoid.

Reply:
No, it doesn't require rust to generate a prologue, it requires the ASM author to know about the calling convention (which is the case for extern C functions) and properly handle it themselves.

`drop` calls are a good point. I guess arguments should be !Drop?

Reply:
@roblabla: What if the calling convention requires generating a prologue? I don't think a naked fn should cause Rust to generate some weird subset of the normal ASM for a function.

Note that it's not sufficient to require arguments to be `!Drop` - they also must not have any drop glue (e.g. `struct MyStruct(String)` is `!Drop` but has drop glue).

I really don't see passing parameters to naked functions as being useful, or even very well-defined.

Reply:
Then they should be `Copy` - which AFAIU denies the drop glue. This is similar to the restrictions `union` currently have AFAIK (unions may not have items that are Drop, so as an approximation they only allow Copy types).

Passing parameters to naked function that have a known ABI (such as extern "C") is both useful and well-defined (After all, you can define functions with global_asm! that take parameters). People have further shown that they are useful - for instance look at [this comment](https://github.com/rust-lang/rfcs/pull/2774#issuecomment-536114348).

Reply:
The example from earlier in the thread shows a case where passing parameters is both useful and well defined:
```rust
#[naked]
pub unsafe fn stat(buf: *const u8, buf: *const u8) {
    asm!("mov $0x20000bc, %%eax; syscall");
}
```